syntax = "proto3";

package notification;

service NotificationService {
  // Get notifications for a user
  rpc GetUserNotifications(GetUserNotificationsRequest) returns (GetUserNotificationsResponse) {}
  
  // Mark a notification as read
  rpc MarkNotificationAsRead(MarkNotificationAsReadRequest) returns (MarkNotificationAsReadResponse) {}
  
  // Subscribe to notifications for a location
  rpc SubscribeToLocation(SubscribeToLocationRequest) returns (SubscribeToLocationResponse) {}
  
  // Unsubscribe from a location
  rpc UnsubscribeFromLocation(UnsubscribeFromLocationRequest) returns (UnsubscribeFromLocationResponse) {}
  
  // Get user's location subscriptions
  rpc GetUserSubscriptions(GetUserSubscriptionsRequest) returns (GetUserSubscriptionsResponse) {}
  
  // Create notification for post like
  rpc CreatePostLikeNotification(CreatePostLikeNotificationRequest) returns (Notification) {}
  
  // Create notification for post comment
  rpc CreatePostCommentNotification(CreatePostCommentNotificationRequest) returns (Notification) {}
  
  // Create notification for comment like
  rpc CreateCommentLikeNotification(CreateCommentLikeNotificationRequest) returns (Notification) {}
  
  // Create notification for comment reply
  rpc CreateCommentReplyNotification(CreateCommentReplyNotificationRequest) returns (Notification) {}
  
  // Create notification for trending post
  rpc CreateTrendingPostNotification(CreateTrendingPostNotificationRequest) returns (Notification) {}
  
  // Notify users about trending posts in their area
  rpc NotifyTrendingPosts(NotifyTrendingPostsRequest) returns (NotifyTrendingPostsResponse) {}
}

// Request to get user notifications
message GetUserNotificationsRequest {
  int32 user_id = 1;
  int32 page = 2;
  int32 page_size = 3;
  bool unread_only = 4;
}

// Response containing user notifications
message GetUserNotificationsResponse {
  repeated Notification notifications = 1;
  int32 total_count = 2;
  int32 unread_count = 3;
}

// Request to mark notification as read
message MarkNotificationAsReadRequest {
  int32 notification_id = 1;
  int32 user_id = 2;
}

// Response for marking notification as read
message MarkNotificationAsReadResponse {
  bool success = 1;
}

// Request to subscribe to location
message SubscribeToLocationRequest {
  int32 user_id = 1;
  double latitude = 2;
  double longitude = 3;
  double radius_km = 4;
}

// Response containing created subscription
message SubscribeToLocationResponse {
  LocationSubscription subscription = 1;
}

// Request to unsubscribe from location
message UnsubscribeFromLocationRequest {
  int32 subscription_id = 1;
  int32 user_id = 2;
}

// Response for unsubscribing from location
message UnsubscribeFromLocationResponse {
  bool success = 1;
}

// Request to get user subscriptions
message GetUserSubscriptionsRequest {
  int32 user_id = 1;
}

// Response containing user subscriptions
message GetUserSubscriptionsResponse {
  repeated LocationSubscription subscriptions = 1;
}

// Request to create post like notification
message CreatePostLikeNotificationRequest {
  int32 post_id = 1;
  int32 post_owner_id = 2;
  int32 liker_id = 3;
}

// Request to create post comment notification
message CreatePostCommentNotificationRequest {
  int32 post_id = 1;
  int32 post_owner_id = 2;
  int32 commenter_id = 3;
  string comment_text = 4;
}

// Request to create comment like notification
message CreateCommentLikeNotificationRequest {
  int32 comment_id = 1;
  int32 comment_owner_id = 2;
  int32 liker_id = 3;
}

// Request to create comment reply notification
message CreateCommentReplyNotificationRequest {
  int32 comment_id = 1;
  int32 comment_owner_id = 2;
  int32 replier_id = 3;
  string reply_text = 4;
}

// Request to create trending post notification
message CreateTrendingPostNotificationRequest {
  int32 post_id = 1;
  int32 user_id = 2;
  string location_name = 3;
}

// Request to notify about trending posts
message NotifyTrendingPostsRequest {
  int32 post_id = 1;
  double latitude = 2;
  double longitude = 3;
  double radius_km = 4;
  string location_name = 5;
}

// Response for trending posts notification
message NotifyTrendingPostsResponse {
  repeated Notification notifications = 1;
  int32 total_notified = 2;
}

// Notification message
message Notification {
  int32 id = 1;
  int32 user_id = 2;
  string type = 3;
  string message = 4;
  int32 entity_id = 5;
  string entity_type = 6;
  bool is_read = 7;
  string created_at = 8;
  string context = 9;
}

// Location subscription message
message LocationSubscription {
  int32 id = 1;
  int32 user_id = 2;
  double latitude = 3;
  double longitude = 4;
  double radius_km = 5;
  string created_at = 6;
  bool is_active = 7;
} 